src/toil/provisioners/abstractProvisioner.py

```diff
@@ -242,6 +244,95 @@ def destroyCluster(self):
+
+    @staticmethod
+    def columnNames(asStr=False):
+        """
+        Return the column headers used in the file tracking information about launched clusters.
+
+        :param asStr: a bool indicating if the return type should be a list or string
+        :returns: The column headers as a list or string
+        """
+        if asStr:
+            return 'clustername,provisioner,zone,type,created,status,appliance\n'
+        else:
+            return ['clustername', 'provisioner', 'zone', 'type', 'created', 'status', 'appliance']
+
+    def updateStatusInList(self, status, provisioner):
+        """Update the status of an entry in ~/toilClusterList.json"""
+
+        with open(self.clusterListPath(), 'r') as f:
+            clusters = json.load(f)
+
+        clusters[provisioner][self._zone][self.clusterName]['status'] = status
+
+        with open(self.clusterListPath(), 'w') as f:
+            json.dump(clusters, f)
+
+    def createJSONStructure(self, path):
+        """
+        Create the template for the json storing information about existing clusters.
+
+        The structure is as follows:
+            { Provisioner:
+                { Zone :
+                    { clusterName:
+                        { 'Cluster': 'Info' }
+                    }
+                }
+            }
+
+        This function only creates the first two levels to facilitate adding entries in
+        AbstractProvisioner.addClusterToList.
+        """
+
+        skeleton = {'aws': dict(),
+                    'gce': dict(),
+                    'azure': dict()}
+
+        with open(path, 'w') as f:
+            json.dump(skeleton, f)
+
+    def addClusterToList(self, provisioner, instanceType):
+        """Save information about launched clusters to a local file to be displayed later."""
+        clusterList = self.clusterListPath()
+
+        # If the file doesnt exist yet, write the skeleton.
+        if not os.path.exists(clusterList):
+            self.createJSONStructure(clusterList)
+
+        with open(clusterList, 'r') as f:
+            clusters = json.load(f)
+
+        newEntry = {self.clusterName: {'instanceType': instanceType,
+                                       'created': time.strftime("%Y-%m-%d %H:%M"),
+                                       'status': 'initializing',
+                                       'appliance': applianceSelf()}}
+
+        clusters[provisioner][self._zone] = newEntry
+
+        with open(clusterList, 'w') as f:
+            json.dump(clusters, f)
+
+    @staticmethod
+    def removeClusterFromList(name, provisioner, zone):
+        """Remove a given cluster's information from the list of active clusters."""
+        listPath = AbstractProvisioner.clusterListPath()
+
+        if os.path.exists(listPath):
+            with open(listPath, 'r') as f:
+                clusters = json.load(f)
+
+            if clusters[provisioner][zone][name]:
```

Comment:
This needs to be: `if name in clusters[provisioner][zone]:` otherwise it will still produce a `KeyError`.

Latest tests:
```
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/venv/bin/toil", line 9, in <module>
    load_entry_point('toil', 'console_scripts', 'toil')()
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/utils/toilMain.py", line 35, in main
    module.main()
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/utils/toilLaunchCluster.py", line 171, in main
    azureStorageCredentials=config.azureStorageCredentials)
  File "/usr/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/utils/toilLaunchCluster.py", line 34, in updatStatusInList
    cluster.updateStatusInList('broken', provisioner)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/provisioners/abstractProvisioner.py", line 271, in updateStatusInList
    clusters[provisioner][self._zone][self.clusterName]['status'] = status
KeyError: 'provisioner-test-d1ca6e2f-bb93-4ac0-8f4b-c58314833937'
INFO:toil.provisioners.aws.awsProvisioner:Terminating instance(s): [u'i-09e5df59dfa140482']
INFO:toil.provisioners.aws.awsProvisioner:Instance(s) terminated.
INFO:toil.provisioners.aws.awsProvisioner:... Waiting for instance(s) to shut down...
INFO:toil.provisioners.aws.awsProvisioner:Instance(s) terminated.
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/venv/bin/toil", line 9, in <module>
    load_entry_point('toil', 'console_scripts', 'toil')()
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/utils/toilMain.py", line 35, in main
    module.main()
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/utils/toilDestroyCluster.py", line 29, in main
    cluster.destroyCluster()
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/provisioners/aws/awsProvisioner.py", line 233, in destroyCluster
    self.removeClusterFromList(name=self.clusterName, provisioner='aws', zone=self._zone)
  File "/mnt/ephemeral/workspace/toil-pull-request-integration-tests/src/toil/provisioners/abstractProvisioner.py", line 330, in removeClusterFromList
    if clusters[provisioner][zone][name]:
KeyError: 'provisioner-test-d1ca6e2f-bb93-4ac0-8f4b-c58314833937'
```
http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/2802/testReport/junit/src.toil.test.provisioners.aws.awsProvisionerTest/AWSStaticAutoscaleTest/testAutoScale_2/